pkg/controller/batchrelease/workloads/statefulset_like_controller.go

```diff
@@ -0,0 +1,179 @@
+		}
+	}
+	return c.statefulSet, nil
+}
+
+func (c *StatefulSetLikeController) GetWorkloadInfo() (*util.WorkloadInfo, error) {
+	set, err := c.GetWorkloadObject()
+	if err != nil {
+		return nil, err
+	}
+
+	workloadInfo := util.ParseWorkloadInfo(set, c.namespacedName)
+	if util.ParseInt32PartitionFrom(set) == *workloadInfo.Replicas {
+		workloadInfo.Paused = true
+	}
+	return workloadInfo, nil
+}
+
+func (c *StatefulSetLikeController) ClaimWorkload() (bool, error) {
+	set, err := c.GetWorkloadObject()
+	if err != nil {
+		return false, err
+	}
+
+	err = util.ClaimWorkload(c.Client, c.planController, set, map[string]interface{}{
+		"rollingUpdate": map[string]interface{}{
+			"partition": pointer.Int32(util.ParseReplicasFrom(set)),
+		},
+	})
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(3).Infof("Claim StatefulSet(%v) Successfully", c.namespacedName)
+	return true, nil
+}
+
+func (c *StatefulSetLikeController) ReleaseWorkload(cleanup bool) (bool, error) {
+	set, err := c.GetWorkloadObject()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	err = util.ReleaseWorkload(c.Client, set)
+	if err != nil {
+		c.recorder.Eventf(c.planController, v1.EventTypeWarning, "ReleaseCloneSetFailed", err.Error())
+		return false, err
+	}
+
+	klog.V(3).Infof("Release StatefulSet(%v) Successfully", c.namespacedName)
+	return true, nil
+}
+
+func (c *StatefulSetLikeController) UpgradeBatch(canaryReplicasGoal, stableReplicasGoal int32) (bool, error) {
+	set, err := c.GetWorkloadObject()
+	if err != nil {
+		return false, err
+	}
+
+	observedReplicas := canaryReplicasGoal + stableReplicasGoal
+	if observedReplicas != util.ParseReplicasFrom(set) {
+		return false, fmt.Errorf("StatefulSet(%v) scaled, should handle scale event first", c.namespacedName)
+	}
+
+	err = util.PatchUpdateStrategy(c.Client, set, map[string]interface{}{
+		"rollingUpdate": map[string]interface{}{
+			"partition": pointer.Int32(stableReplicasGoal),
+		},
+	})
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(3).Infof("Upgrade StatefulSet(%v) Partition to %v Successfully", c.namespacedName, stableReplicasGoal)
+	return true, nil
+}
+
+func (c *StatefulSetLikeController) IsBatchUpgraded(canaryReplicasGoal, stableReplicasGoal int32) (bool, error) {
+	set, err := c.GetWorkloadObject()
+	if err != nil {
+		return false, err
+	}
+
+	if !util.IsRollingUpdateStrategy(set) {
+		return false, fmt.Errorf("StatefulSet(%v) rollingUpdate configuration is nil, should check it maully", c.namespacedName)
+	}
+
+	partition := util.ParseInt32PartitionFrom(set)
+	return partition <= stableReplicasGoal, nil
+}
+
+func (c *StatefulSetLikeController) IsBatchReady(canaryReplicasGoal, stableReplicasGoal int32) (bool, error) {
+	workloadInfo, err := c.GetWorkloadInfo()
+	if err != nil {
+		return false, err
+	}
+
+	// make sure all pod is available, and the canary goal is met
+	ready := workloadInfo.Status.ReadyReplicas == workloadInfo.Status.Replicas &&
```

Comment:
condition workloadInfo.Status.ReadyReplicas == workloadInfo.Status.Replicas is too strict, there is maybe cases where old replicas are not ready